# Máquinas de Touring

<img src = "https://www.ellitoral.com/images/2023/05/11/Klmqnm6CD_1300x655__1.jpg"
   style="display: block; margin-left: auto; margin-right: auto;"/>

## Desglose del código

Se define la clase **MaquinaDeTouring**
Y se inserta el constructor **__init__** en el que se definen las variables que contendrá nuestro código:

**self.cinta:** La cinta de la máquina de Turing. Es una lista de caracteres, cada uno representando un símbolo en la cinta.

**self.posicion_del_cabezal:** La posición del cabezal de lectura/escritura. Comienza en 0 (la primera posición de la cinta).

**self.estado_actual:** El estado actual de la máquina. Comienza en el estado inicial proporcionado.

**self.estados_de_aceptacion:** Un conjunto de estados que indican cuando la máquina ha terminado de procesar (estados de aceptación).

**self.funcion_de_transicion:** Un diccionario que define las reglas de transición de la máquina.

In [13]:
class MaquinaDeTuring:
    def __init__(self, cinta, estado_inicial, estados_de_aceptacion, funcion_de_transicion):
        self.cinta = list(cinta)
        self.posicion_del_cabezal = 0
        self.estado_actual = estado_inicial
        self.estados_de_aceptacion = estados_de_aceptacion
        self.funcion_de_transicion = funcion_de_transicion

Se define el método **Paso**

In [14]:
def paso(self):
    simbolo = self.cinta[self.posicion_del_cabezal]
    accion = self.funcion_de_transicion.get((self.estado_actual, simbolo))

En donde:

**simbolo:** Obtiene el símbolo bajo el cabezal de lectura/escritura.

**accion:** Busca en la función de transición la acción correspondiente al par (estado actual, símbolo). Si no se encuentra ninguna acción, se retorna False y la máquina se detiene.

Si accion es None se desempaqueta la acción obtenida. La acción incluye el nuevo estado, el nuevo símbolo que se escribirá en la cinta, y la dirección en la que se moverá el cabezal (D para derecha, I para izquierda).

Se actualiza la cinta y se actualiza el estado.

In [16]:
if accion is None:
        return False

    nuevo_estado, nuevo_simbolo, direccion = accion
    self.cinta[self.posicion_del_cabezal] = nuevo_simbolo
    self.estado_actual = nuevo_estado

IndentationError: unindent does not match any outer indentation level (<string>, line 4)

Luego de moverá el cabezal a la derecha o a la izquierda según la dirección especificada. Si el cabezal se mueve fuera de los límites de la cinta, se extiende la cinta con un símbolo en blanco (_).

In [ ]:
if direccion == 'D':
        self.posicion_del_cabezal += 1
        if self.posicion_del_cabezal >= len(self.cinta):
            self.cinta.append('_')  # Extiende la cinta con un símbolo en blanco
    elif direccion == 'I':
        if self.posicion_del_cabezal > 0:
            self.posicion_del_cabezal -= 1
        else:
            self.cinta.insert(0, '_')  # Extiende la cinta con un símbolo en blanco
    return True

Se define el método **ejecutar** en donde se realizará un bucle en donde se ejecutaran los pasos hasta que la máquina alcanza un estado de aceptación.

Luego se realiza un paso de la máquina. Si paso retorna False, significa que no hay más transiciones posibles y el bucle se detiene.

Y finalmente devuelve el contenido de la cinta como una cadena de caracteres.

In [ ]:
def ejecutar(self):
    while self.estado_actual not in self.estados_de_aceptacion:
        if not self.paso():
            break
    return ''.join(self.cinta)

## Código completo y caso de uso

In [ ]:
class MaquinaDeTuring:
    def __init__(self, cinta, estado_inicial, estados_de_aceptacion, funcion_de_transicion):
        self.cinta = list(cinta)
        self.posicion_del_cabezal = 0
        self.estado_actual = estado_inicial
        self.estados_de_aceptacion = estados_de_aceptacion
        self.funcion_de_transicion = funcion_de_transicion

    def paso(self):
        simbolo = self.cinta[self.posicion_del_cabezal]
        accion = self.funcion_de_transicion.get((self.estado_actual, simbolo))
        
        if accion is None:
            return False
        
        nuevo_estado, nuevo_simbolo, direccion = accion
        self.cinta[self.posicion_del_cabezal] = nuevo_simbolo
        self.estado_actual = nuevo_estado
        
        if direccion == 'D':
            self.posicion_del_cabezal += 1
            if self.posicion_del_cabezal >= len(self.cinta):
                self.cinta.append('_')  # Extiende la cinta con un símbolo en blanco
        elif direccion == 'I':
            if self.posicion_del_cabezal > 0:
                self.posicion_del_cabezal -= 1
            else:
                self.cinta.insert(0, '_')  # Extiende la cinta con un símbolo en blanco
        return True

    def ejecutar(self):
        while self.estado_actual not in self.estados_de_aceptacion:
            if not self.paso():
                break
        return ''.join(self.cinta)
    
cinta = "0110"
estado_inicial = "q0"
estados_de_aceptacion = {"q_aceptar"}
funcion_de_transicion = {
    ("q0", "0"): ("q0", "1", "D"),
    ("q0", "1"): ("q0", "0", "D"),
    ("q0", "_"): ("q_aceptar", "_", "D"),
}

mt = MaquinaDeTuring(cinta, estado_inicial, estados_de_aceptacion, funcion_de_transicion)

resultado = mt.ejecutar()
print(f"Resultado final en la cinta: {resultado}")

Resultado final en la cinta: 1001__


La máquina comienza en el estado "q0" y lee el primer símbolo "0".
Según la función de transición, "0" se convierte en "1" y el cabezal se mueve a la derecha.
La máquina repite este proceso para cada símbolo en la cinta.
Cuando el cabezal alcanza el final de la cinta (representado por "_"), la máquina cambia al estado de aceptación "q_aceptar" y se detiene.
El contenido final de la cinta se muestra como "1001".